iceoryx_utils/include/iceoryx_utils/concurrent/resizeable_lockfree_queue.hpp

```diff
@@ -0,0 +1,198 @@
+
+    // multiple overloads to set the capacity
+    // 1) The most general one allows providing a removeHandler to specify remove behavior.
+    // 2) The overload where a container is provided can be used if the removed elements are to be stored.
+    //    The container must satisfy certain requirements, such as providing push_back.
+    // 3) The final overload discards removed elements.
+
+    /// @brief      Set the capacity to some value.
+    /// @param[in]  newCapacity capacity to be set
+    /// @param[in]  removeHandler is a function taking an element which specifies
+    ///             what to do with removed elements should the need for removal arise.
+    /// @return     true if the capacity was successfully set, false otherwise
+    template <typename Function,
+              typename = typename std::enable_if<cxx::is_invocable<Function, ElementType>::value>::type>
+    bool setCapacity(uint64_t newCapacity, Function&& removeHandler) noexcept;
+
+    /// @brief Set the capacity to a new capacity between 0 and MaxCapacity, if the capacity is reduced
+    /// it may be necessary to remove the least recent elements.
+    /// @param[in] newCapacity new capacity to be set, if it is larger than MaxCapacity the call fails
+    /// @param[out] removedElements container were potentially removed elements can be stored.
+    /// @return true setting if the new capacity was successful, false otherwise (newCapacity > MaxCapacity)
+    /// @note threadsafe, lockfree but multiple concurrent calls may have no effect
+    template <typename ContainerType = iox::cxx::vector<ElementType, MaxCapacity>,
+              typename = typename std::enable_if<!cxx::is_invocable<ContainerType, ElementType>::value>::type>
+    bool setCapacity(uint64_t newCapacity, ContainerType& removedElements) noexcept;
+
+    /// @brief Set the capacity to a new capacity between 0 and MaxCapacity, if the capacity is reduced
+    /// it may be necessary to remove the least recent elements which are then discarded.
+    /// @param[in] newCapacity new capacity to be set, if it is larger than MaxCapacity the call fails
+    /// @return true setting if the new capacity was successful, false otherwise (newCapacity > MaxCapacity)
+    /// @note threadsafe, lockfree but multiple concurrent calls may have no effect
+    bool setCapacity(uint64_t newCapacity) noexcept;
+
+  private:
+    using BufferIndex = typename Base::BufferIndex;
+    std::atomic<uint64_t> m_capacity{MaxCapacity};
+
+    // we also sync m_capacity with this flag
+    std::atomic_flag m_resizeInProgress{false};
+
+    // Remark: The vector m_unusedIndices is protected by the atomic flag, but this also means dying during a resize
+    // will prevent further resizes (This is not a problem for the use case were only the dying receiver itself requests
+    // the resize.)
+    // I.e. resize is lockfree, but not in a useful and robust way as it assumes that a concurrent resize will
```

Comment:
Change comment. It is self, but with the stated limitations  which are acceptable for our use case since we will not have concurrent resizes and they make not too much sense because the result is unpredictable (it will be one of the requested new capacities).

Reply:
That is not exactly true, it can be set by whoever wants to do so, but will interfere with other concurrent setCapacity operations. But this is the same for other concurrent operations as well, a concurrent pop will affect what another pop gets. This is just how things are in concurrent programming.

But yes, I can emphasize this more.

Reply:
maybe I expressed it poorly. I didn't meant that the resize operation is unsafe when done concurrently, but in case the application dies during the resize, the queue is broken. This is a non issue if that application is also the owner of the queue.

Reply:
No, I got what you were saying. I just wanted to point out that stating the limitations correctly is a bit tricky. Added a note to do so.